[DOC-444] [2.23] Lightweight DB clone (#23515)
yugabyte/yugabyte-db: docs/content/preview/manage/backup-restore/lightweight-db-clone.md

    production_db=# DROP TABLE t1;
    ```

    ```output
    DROP TABLE
    ```

    ```sql
    production_db=# SELECT * FROM t1 ORDER BY k;
    ```

    ```output
    ERROR:  relation "t1" does not exist
    LINE 1: SELECT * FROM t1 ORDER BY k;
    ```

   The table is dropped and there is no way you can query it.

1. Create a database `clone_db` using `production_db` as the template and using the timestamp generated in step 4.

    ```sql
    production_db=# CREATE DATABASE clone_db TEMPLATE production_DB AS OF 1723243720285350;
    ```

    ```sql
    \c clone_db
    ```

    ```output
    You are now connected to database "clone_db" as user "yugabyte".
    ```

    ```sql
    clone_db=# SELECT * FROM t1 ORDER BY k;
    ```

    ```output
     k | v
    ---+---
     1 | 1
     2 | 0
     3 | 1
     4 | 0
     5 | 1
    (5 rows)
    ```

    `clone_db` contains all the data from `production_db` at the specified timestamp, which means you can read table `t1` that was dropped accidentally. You now have two isolated databases that can serve reads and writes independently. You can copy the data back to `production_db` or switch the workload to `clone_db`.

## Best practices

> Although creating a clone database is quick and doesn't take up much disk space as no data is copied, it's important to understand that a clone does create an independent set of logical tablets. This effectively doubles the number of tablets, although the clone database tablets share the same data files as the original database. Keep in mind the following impacts:

Reviewer: ```suggestion
Although creating a clone database is quick and initially doesn't take up much added disk space as no data is copied, it's important to understand that a clone does create an independent set of logical tablets. This effectively doubles the number of tablets, although the clone database tablets share the same data files as the original database. Keep in mind the following impacts:
```

---
title: Database clone
headerTitle: Database clone
linkTitle: Database clone
description: Clone your database in YugabyteDB for data recovery, development, and testing.
badges: tp
menu:
  preview:
    identifier: lightweight-db-clone
    parent: backup-restore
    weight: 706
type: docs
---

Instant database cloning in YugabyteDB allows you to quickly create a zero-copy, independent writable clone of your database that can be used for data recovery, development, and testing. Cloning is both fast and efficient because when initially created, it shares the same data files with the original database. Subsequently, as data is written to the clone, the clone stores its own changes as separate and independent delta files. Although they physically share some files, the two databases are logically isolated, which means you can freely play with the clone database, perform DDLs, read and write data, and delete it without affecting the original database.

> You can create clones as of now, or as of any time in the recent past, within a configurable history retention period. This is particularly useful for data recovery due to user or application error scenarios. For instance, if you dropped a table by mistake at 9:01, you can create a clone of the database at 9:00 (before the table drop) and restore the lost data back to the production database.

Reviewer 2: I suggest the following extension to clarify the value of cloning in relation to data recovery.
```suggestion
You can create clones as of now, or as of any time in the recent past, within a configurable history retention period. This is particularly useful for data recovery due to user or application error scenarios. For instance, if you dropped a table by mistake at 9:01, then you detected this error at 10:45, you are interested in getting the lost data as it was at 9:00 (before the drop table). However, you don't want to lose the data inserted on other tables between 9:01 and 10:45.  With DB cloning, you can create a clone of the database at 9:00 (before the table drop) and restore the lost data back to the production database.
```

        "restore_time": "2024-08-09 21:42:55.048663"
    }
]
```

You can find the `source_database_id` or `source_namespace_id` from the [YB-Master leader UI](../../../reference/configuration/default-ports/#servers) under the `/namespaces` end point.

In this example, two clones were made of the source database `00004000000030008000000000000000` that are COMPLETE. The two clones are `testing_clone_db` and `dev_clone_db` and they each have a unique `seq_no` used to identify each clone operation from the same source database.

You can check the status of a specific clone operation if you have both the `source_database_id` (YSQL) or `source_namespace_id`(YCQL) and the `seq_no` as follows:

```sh
./bin/yb-admin --master_addresses $MASTERS list_clones 00004000000030008000000000000000 2
```

```output
[
    {
        "aggregate_state": "COMPLETE",
        "source_namespace_id": "00004000000030008000000000000000",
        "seq_no": "2",
        "target_namespace_name": "dev_clone_db",
        "restore_time": "2024-08-09 21:42:55.048663"
    }
]
```

Use the `list_clones` command to check whether a clone operation completed successfully or not.

> Note that the cluster doesn't allow two clone operations to happen concurrently on the same source database.

Reviewer 2: giving this another look, I think we should make it clearer as it caused some confusion internally so we should be clearer with our public-facing docs. In this regard, I suggest:
```suggestion
Note that the cluster doesn't allow you to perform two clone operations concurrently on the same source database. You have to wait for the first clone to finish until you can perform another clone. However, you can create as many clones as you want for the same database or different databases.
```

### Example

{{% explore-setup-single-local %}}

The following example uses [ysqlsh](../../../admin/ysqlsh/) to create a database clone to recover from an accidental table deletion.

1. Start ysqlsh and create the database as follows:

    ```sh
    ./bin/ysqlsh
    CREATE DATABASE production_db;
    ```

1. Create a snapshot schedule that produces a snapshot once a day (every 1,440 minutes), and retains it for three days (4,320 minutes) using the following command:

    ```sh
    ./bin/yb-admin --master_addresses ip1:7100,ip2:7100,ip3:7100 create_snapshot_schedule 1440 4320 ysql.production_db
    ```

> 1. Create a table `t1` and add some data as follows:

Reviewer 2: ```suggestion
1. Create two tables `t1` and `t2`, and add some data as follows:
```

    ```sql
    \c production_db;
    production_db=# CREATE TABLE t1 (k INT, v INT);
    production_db=# INSERT INTO t1 (k,v) SELECT i,i%2 FROM generate_series(1,5) AS i;
    production_db=# SELECT * FROM t1 ORDER BY k;
    ```

    ```output
    k | v
   ---+---
    1 | 1
    2 | 0
    3 | 1
    4 | 0
    5 | 1
   (5 rows)
    ```

> 

Reviewer 2: Create table t2:
```sql
production_db=# CREATE TABLE t2 (key INT, c1 TEXT);
production_db=# INSERT INTO t2 (key,c1) SELECT i,md5(random()::text) FROM generate_series(1,5) AS i;
production_db=# SELECT * FROM t2 ORDER BY key;
```
```output
 key |                c1                
-----+----------------------------------
   1 | 450e6c49f86c76d944375e29e48f2dee
   2 | b934a3bdf7438458a85b0858c41f731c
   3 | 08697ed89ec387e714c6587e522d7a7e
   4 | a879ff99872b3c3433803d3c3229f0cf
   5 | 4d46a53780a7a348179e1af9b692e95e
(5 rows)
```

1. Determine the exact time when your database is in the correct state. This timestamp will be used to create a clone of the production database from the point when it was in the desired state. Execute the following SQL query to retrieve the current time in UNIX timestamp format:

    ```sql
    production_db=# SELECT (EXTRACT (EPOCH FROM CURRENT_TIMESTAMP)*1000000)::decimal(38,0);
    ```

    ```output
         numeric
     ------------------
      1723243720285350
      (1 row)
    ```

1. Drop the table `t1` simulating a user error dropping the table accidentally.

    ```sql
    production_db=# DROP TABLE t1;
    ```

    ```output
    DROP TABLE
    ```

> 

Reviewer 2: We can augment the example as follows:

Meanwhile, table t2 is still accepting reads/writes. We can simulate that by inserting 2 more rows as follows:
```sql
INSERT INTO t2 (key,c1) SELECT i,md5(random()::text) FROM generate_series(6,7) AS i;
```
After that, you try to query table `t1`. However, you figure out the table is dropped and there is no way you can query it.

    ```sql
    production_db=# SELECT * FROM t1 ORDER BY k;
    ```

    ```output
    ERROR:  relation "t1" does not exist
    LINE 1: SELECT * FROM t1 ORDER BY k;
    ```

   The table is dropped and there is no way you can query it.

> 1. Create a database `clone_db` using `production_db` as the template and using the timestamp generated in step 4.

Reviewer 2: ```suggestion
1. To get back lost data, create a database `clone_db` using `production_db` as the template and using the timestamp generated in step 4.
```

    ```sql
    production_db=# CREATE DATABASE clone_db TEMPLATE production_DB AS OF 1723243720285350;
    ```

    ```sql
    \c clone_db
    ```

    ```output
    You are now connected to database "clone_db" as user "yugabyte".
    ```

    ```sql
    clone_db=# SELECT * FROM t1 ORDER BY k;
    ```

    ```output
     k | v
    ---+---
     1 | 1
     2 | 0
     3 | 1
     4 | 0
     5 | 1
    (5 rows)
    ```

>     `clone_db` contains all the data from `production_db` at the specified timestamp, which means you can read table `t1` that was dropped accidentally. You now have two isolated databases that can serve reads and writes independently. You can copy the data back to `production_db` or switch the workload to `clone_db`.

Reviewer 2: ```suggestion
    `clone_db` contains all the data from `production_db` at the specified timestamp, which means you can read table `t1` that was dropped accidentally. You can copy the lost data from table t1 back to `production_db` by exporting the data from the clone and importing it into `production_db`. Alternatively, you can switch the workload to `clone_db`. You now have two isolated databases that can serve reads and writes independently.
```

## Best practices

Although creating a clone database is quick and initially doesn't take up much added disk space as no data is copied, it's important to understand that a clone does create an independent set of logical tablets. This effectively doubles the number of tablets, although the clone database tablets share the same data files as the original database. Keep in mind the following impacts:

- Higher CPU usage due to the additional tablets
- Increased memory consumption from the extra tablets

> - Potentially up to twice the disk usage (after major compactions) due to the retention of older data versions

Reviewer 2: ```suggestion
- Increased space usage only if compaction is performed on clone or original database. This is because both original and post-compaction data files are kept in disk i.e.: say compaction is performed on the source database which generates new compacted files. The old data files are kept to serve reads for the clone database.
```

---
title: Database clone
headerTitle: Database clone
linkTitle: Database clone
description: Clone your database in YugabyteDB for data recovery, development, and testing.
badges: tp
menu:
  preview:
    identifier: lightweight-db-clone
    parent: backup-restore
    weight: 706
type: docs
---

> Database cloning in YugabyteDB allows you to quickly create an independent copy of your database for data recovery, development, and testing. The process is efficient because the clone starts by using the same data files as the original database. After the initial creation, the cloned database serves reads and writes independently, creates and maintains its own changes as delta files, separately from the original database.

Reviewer 2: What are the objections to the `data recovery` word? What I meant is that it can be used to restore deleted data by creating a clone as of a moment in the past, doing forensic analysis to export the deleted data from the clone and import it back to the original database.

> Database cloning in YugabyteDB allows you to quickly create an independent copy of your database for data recovery, development, and testing. The process is efficient because the clone starts by using the same data files as the original database. After the initial creation, the cloned database serves reads and writes independently, creates and maintains its own changes as delta files, separately from the original database.

Reviewer: Sure we can go with data recovery. See my later comment though about why wouldn't I just use PITR? Maybe a brief mention earlier in use case describing what you wrote here in your comment, this flow wasn't clear to me and is more cumbersome than PITR; so maybe also add small description of when/why use this instead of vanilla PITR

The two databases are completely isolated, meaning you can experiment with the cloned database, perform DDL operations, read and write data, and even delete the clone without impacting the original database.

Additionally, you can create a clone of the database from a specific point in time within a configurable retention period, making it particularly useful in data recovery scenarios due to user or application error. For example, if you accidentally dropped a table at 9:01, you can create a clone from 9:00 (just before the table drop) to restore the lost data to the production database.

> 

Reviewer: You can't have more than one clone of a database at the same time.

## Use cases

### Data recovery

If you encounter data loss due to user error (for example, accidentally dropping a table) or application error (for example, updating rows with corrupted data), you can quickly create a clone of your production database from a point in time when database was in a good state. This allows you to perform forensic analysis and restore the lost or corrupted data back to the original database.

### Development and testing

You can instantly create an isolated copy of your production database for development and testing purposes. This copy includes the latest data from the production database, enabling developers to test their changes on an identical copy of the production database without affecting the performance of the production database.

## Enable database cloning

To enable lightweight database cloning in a cluster, you set the YB-Master flag `enable_db_clone` to true. Additionally, as cloning is in {{<badge/tp>}}, the flag `enable_db_clone` should be added to the `allowed_preview_flags_csv`list.

> **Using yb-ctl**: You can set these two flags using the `--master_flag` option of [yb-ctl](../../../admin/yb-ctl/) as follows:

Reviewer: yugabyted from now on

```sh
./bin/yugabyted start --master_flags "allowed_preview_flags_csv={enable_db_clone},enable_db_clone=true"
```

**Using yb-ts-cli**: You can set the runtime flags while the yb-master process is running using [set_flag](../../../admin/yb-ts-cli/#set-flag) option of [yb-ts-cli](../../../admin/yb-ts-cli/) as follows:

```sh
./bin/yb-ts-cli --server-address=master_host:7100 set_flag allowed_preview_flags_csv enable_db_clone

> ./build/latest/bin/yb-ts-cli --server-address=127.0.0.1:7100 set_flag enable_db_clone true

Author: @yamen-haddad - can you clarify regarding this command?

```

> ## How to use database cloning

Reviewer: ```suggestion
## Clone databases
```

### Prerequisites

- [Create a snapshot schedule](../../../manage/backup-restore/point-in-time-recovery/#create-a-schedule) for the database you want to clone.

> This keeps the history of the updates to a certain retention period. After that you can create a clone of the original database as of any point in time within the specified retention period.

Reviewer: ```suggestion
    This keeps the history of updates for the specified retention period. You can create a clone of the database as of any point in time within the specified retention period.
```

> - You have to trust local YSQL connections (that use UNIX domain sockets) in the [host based authentication](../../../secure/authentication/host-based-authentication/). You have to do this for all the yb-tservers in the cluster. You can do this when starting the yb-tserver process by adding the authentication line `local all all trust` to the yb-tserver flag, [ysql_hba_conf_csv](../../../reference/configuration/yb-tserver/#ysql-hba-conf-csv).

Reviewer: ```suggestion
- You have to trust local YSQL connections (that use UNIX domain sockets) in the [host-based authentication](../../../secure/authentication/host-based-authentication/). You have to do this for all YB-TServers in the cluster. You can do this when starting the YB-TServer process by adding the authentication line `local all all trust` to the [ysql_hba_conf_csv](../../../reference/configuration/yb-tserver/#ysql-hba-conf-csv) flag.
```

> For example, if you are using yb-ctl you can run the following command:

Reviewer: no yb-ctl examples

also, indent paragraphs under their bullet

```sh
./bin/yb-ctl create --master_flags "enable_db_clone=true,allowed_preview_flags_csv=enable_db_clone" --tserver_flags '"ysql_hba_conf_csv=""host all all 0.0.0.0/0 trust,local all all trust"""'
```

{{<note title="Note">}}

> Ensure to not override your default host based authentication rules when trusting the local connection. You may need to add additional authentication lines to `ysql_hba_conf_csv` based on your specific configuration. For more information, see [host based authentication](../../../secure/authentication/host-based-authentication/).

Reviewer: ```suggestion
Do not override your default host-based authentication rules when trusting the local connection. You may need to add additional authentication lines to `ysql_hba_conf_csv` based on your specific configuration. For more information, see [host based authentication](../../../secure/authentication/host-based-authentication/).
```

{{</note>}}

> ### Cloning in YSQL

Reviewer: ```suggestion
### Clone a YSQL database
```

Because YugabyteDB is PostgreSQL compatible, you can create a database as a clone of another using the `TEMPLATE` SQL option of `CREATE DATABASE` command as follows:

```sql
CREATE DATABASE clone_db TEMPLATE original_db;
```

> `clone_db` is created as a clone of database `original_db` which contains the latest schema and data of `original_db` as of current time.

Reviewer: ```suggestion
In this example, `clone_db` is created as a clone of `original_db`, and contains the latest schema and data of `original_db` as of current time.
```

> Additionally, to create a clone of the original database at a specific point in time in the past (within the history retention period specified when creating the snapshot schedule) you can specify the [Unix timestamp](https://www.unixtimestamp.com/) in microseconds in the `AS OF` option as follows:

Reviewer: ```suggestion

To create a clone of the original database at a specific point in time (within the history retention period specified when creating the snapshot schedule), you can specify the [Unix timestamp](https://www.unixtimestamp.com/) in microseconds using the `AS OF` option as follows:
```

```sql

> CREATE DATABASE clone_db TEMPLATE origginal_db AS OF 1723146703674480;

Reviewer: API docs need updating?

```

> ### Cloning in YCQL

Reviewer: ```suggestion
### Clone a YCQL keyspace
```

> You can create a clone in YCQL using the yb-admin command as follows:

Reviewer: ```suggestion
You can create a clone in YCQL using the yb-admin `clone_namespace` command as follows:
```

```sh
./bin/yb-admin --master_addresses $MASTERS clone_namespace ycql.originaldb1 clonedb2 1715275616599020
```

> `clonedb2` is created as as a clone of the ycql database `originaldb1` as of 1715275616599020 Unix timestamp.

Reviewer: ```suggestion
In this example, `clonedb2` is created as a clone of `originaldb1` as of 1715275616599020 Unix timestamp.
```

### Check the clone status

> You can check the status of clone operations performed on a database with the `source_database_id` (YSQL)/`source_namespace_id`(YCQL) field using the yb-admin command `list_clones` as follows:

Reviewer: ```suggestion
To check the status of clone operations performed on a database, use the yb-admin `list_clones` command and provide the `source_database_id` (YSQL) or `source_namespace_id` (YCQL), as follows:
```

```sh

> ./bin/yb-admin --master_addresses $MASTERS list_clones 00004000000030008000000000000000

Reviewer: how do you find out what the db ID is so that you can use it in this command?

Author: cc @yamen-haddad

```

```output
[
    {
        "aggregate_state": "COMPLETE",
        "source_namespace_id": "00004000000030008000000000000000",
        "seq_no": "1",
        "target_namespace_name": "testing_clone_db",
        "restore_time": "2024-08-09 21:42:16.451974"
    },
    {
        "aggregate_state": "COMPLETE",
        "source_namespace_id": "00004000000030008000000000000000",
        "seq_no": "2",
        "target_namespace_name": "dev_clone_db",
        "restore_time": "2024-08-09 21:42:55.048663"
    }
]
```

> Note that there were two clone operations performed on the source database `00004000000030008000000000000000` that are COMPLETE. The two clones are `testing_clone_db` and `dev_clone_db` and they each have a unique `seq_no` used to identify each clone operation from the same source database.

Reviewer: ```suggestion
In this example, two clones were made of the source database `00004000000030008000000000000000` that are COMPLETE. The two clones are `testing_clone_db` and `dev_clone_db` and they each have a unique `seq_no` used to identify each clone operation from the same source database.
```

> Alternatively, you can check the status of a specific clone operation if you have both the `source_database_id` (YSQL)/`source_namespace_id`(YCQL) and the `seq_no` as follows:

Reviewer: ```suggestion
You can check the status of a specific clone operation if you have both the `source_database_id` (YSQL) or `source_namespace_id`(YCQL) and the `seq_no` as follows:
```

```sh
./bin/yb-admin --master_addresses $MASTERS list_clones 00004000000030008000000000000000 2
```

> ```output

Reviewer: The output doesn't look any different what changed

Author: my bad. had to include the output only with seq_no=2. Updated it.

[
    {
        "aggregate_state": "COMPLETE",
        "source_namespace_id": "00004000000030008000000000000000",
        "seq_no": "1",
        "target_namespace_name": "testing_clone_db",
        "restore_time": "2024-08-09 21:42:16.451974"
    },
    {
        "aggregate_state": "COMPLETE",
        "source_namespace_id": "00004000000030008000000000000000",
        "seq_no": "2",
        "target_namespace_name": "dev_clone_db",
        "restore_time": "2024-08-09 21:42:55.048663"
    }
]
```

> This command can be helpful to check whether a clone operation completed successfully or not.

Reviewer: ```suggestion
Use the `list_clones` command to check whether a clone operation completed successfully or not.
```

> Note that the cluster doesn't allow two clone operations to happen concurrently on the same source database.

Reviewer: moved this limitation up to top

### Example

> {{% explore-setup-single %}}

Reviewer: actually this can only practically be run using local instance. you can't run yb-admin on Aeon clusters for example

> The following example uses [ysqlsh](../../../admin/ysqlsh/) to create a database clone for recovering from an accidental table deletion by a user.

Reviewer: ```suggestion
The following example uses [ysqlsh](../../../admin/ysqlsh/) to create a database clone to recover from an accidental table deletion.
```

    production_db=# INSERT INTO t1 (k,v) SELECT i,i%2 FROM generate_series(1,5) AS i;
    production_db=# SELECT * FROM t1 ORDER BY k;
    ```

    ```output
    k | v
   ---+---
    1 | 1
    2 | 0
    3 | 1
    4 | 0
    5 | 1
   (5 rows)
    ```

1. Determine the exact time when your database is in the correct state. This timestamp will be used to create a clone of the production database from the point when it was in the desired state. Execute the following SQL query to retrieve the current time in UNIX timestamp format:

    ```sql
    production_db=# SELECT (EXTRACT (EPOCH FROM CURRENT_TIMESTAMP)*1000000)::decimal(38,0);
    ```

    ```output
         numeric
     ------------------
      1723243720285350
      (1 row)
    ```

1. Drop the table `t1` simulating a user error dropping the table accidentally.

    ```sql
    production_db=# DROP TABLE t1;
    ```

    ```output
    DROP TABLE
    ```

    ```sql
    production_db=# SELECT * FROM t1 ORDER BY k;
    ```

    ```output
    ERROR:  relation "t1" does not exist
    LINE 1: SELECT * FROM t1 ORDER BY k;
    ```

   The table is dropped and there is no way you can query it.

> 1. Create a database `clone_db` using `production_db` as the clone template and the clone happens as of the time marked at step 4.

Reviewer: ```suggestion
1. Create a database `clone_db` using `production_db` as the template and using the timestamp generated in step 4.
```

    ```sql
    production_db=# CREATE DATABASE clone_db TEMPLATE production_DB AS OF 1723243720285350;
    ```

    ```sql
    \c clone_db
    ```

    ```output
    You are now connected to database "clone_db" as user "yugabyte".
    ```

    ```sql
    clone_db=# SELECT * FROM t1 ORDER BY k;
    ```

    ```output
     k | v
    ---+---
     1 | 1
     2 | 0
     3 | 1
     4 | 0
     5 | 1
    (5 rows)
    ```

>     `clone_db` is created and it contains all the data of `production_db` as of time at step 4 which means you can read table `t1` that was dropped accidentally. You can copy the data back to `production_db` or switch the workload to `clone_db`. You now have two isolated databases that can serve read and write independently.

Reviewer: ```suggestion
    `clone_db` is contains all the data from `production_db` at the specified timestamp, which means you can read table `t1` that was dropped accidentally. You can copy the data back to `production_db` or switch the workload to `clone_db`. You now have two isolated databases that can serve reads and writes independently.
```

## Best practices

> Although creating a clone database is quick and doesn't have any space amplification as there is no physical data copying, it's important to understand that a clone creates an independent set of logical tablets. This effectively doubles the number of tablets, although the clone database's tablets share the same data files as the original database. Keep in mind the following impacts:

Reviewer: ```suggestion
Although creating a clone database is quick and doesn't take up much disk space as no data is copied, it's important to understand that a clone does create an independent set of logical tablets. This effectively doubles the number of tablets, although the clone database tablets share the same data files as the original database. Keep in mind the following impacts:
```

- Higher CPU usage due to the additional tablets
- Increased memory consumption from the extra tablets

> - Potentially up to twice the disk usage (after major compactions) due to the retention of older data versions

Reviewer: Not sure I understand - above you say there is no 'space amplification' - does that refer to disk space or not?

---
title: Database clone
headerTitle: Database clone

> linkTitle: Database clone

Reviewer 2: Replace the feature name to: Instant database cloning

description: Clone your database in YugabyteDB for data recovery, development, and testing.
badges: tp
menu:
  preview:

>     identifier: lightweight-db-clone

Reviewer 2: ```suggestion
    identifier: instant-db-clone
```
